From ec179e3e721663513ada30f1776b21468e23d50d Mon Sep 17 00:00:00 2001
From: Benoit DOLEZ <bdolez@ant-computing.com>
Date: Tue, 4 Feb 2020 09:28:48 +0100
Subject: [PATCH] parseNameValue: fix no quoting support

Function description announce support of quoted variable :
- name=value
- name="value"
- name='value'

... but there is nothing in the code to really support quoted string.
I had small fixes to support it.

My tests :
$ cat > test.txt <<EOF
a=
a=123
a=123
a='123'
a="123"
a=123 b=456
a=123 b="456"
a=123 b="456" c=
a=123 b="456'789" c=
a=123 b="456 789" c=
a=123 b="456 789' c=
a=123 b=456 789 c=
EOF

$ cat > test.rb << EOF
version=2
rule=test-ok:%[
  {"type": "name-value-list"}
  ]%
EOF

$ cat test.txt | ./lognormalizer -r test.rb -T
{ "event.tags": [ "test-ok" ] }
{ "event.tags": [ "test-ok" ] }
{ "event.tags": [ "test-ok" ] }
{ "event.tags": [ "test-ok" ] }
{ "event.tags": [ "test-ok" ] }
{ "event.tags": [ "test-ok" ] }
{ "event.tags": [ "test-ok" ] }
{ "event.tags": [ "test-ok" ] }
{ "event.tags": [ "test-ok" ] }
{ "event.tags": [ "test-ok" ] }
{ "originalmsg": "a=123 b=\"456 789' c= ", "unparsed-data": "a=123 b=\"456 789' c= " }
{ "originalmsg": "a=123 b=456 789 c= ", "unparsed-data": "a=123 b=456 789 c= " }
---
 src/parser.c | 14 +++++++++++++-
 1 file changed, 13 insertions(+), 1 deletion(-)

diff --git a/src/parser.c b/src/parser.c
index 7343d84..8253791 100644
--- a/src/parser.c
+++ b/src/parser.c
@@ -2420,11 +2420,23 @@ parseNameValue(npb_t *const npb,
 	const size_t lenName = i - iName;
 	++i; /* skip '=' */
 
+	char quoting = npb->str[i]; // end of string
+	if (i < npb->strLen && (quoting == '"' || quoting == '\''))
+		++i;
+	else
+		quoting = 0; // str[i] can't be null, is a good default value
+
 	const size_t iVal = i;
-	while(i < npb->strLen && !isspace(npb->str[i]))
+	while(i < npb->strLen &&
+	     ((quoting && npb->str[i] != quoting) || (!quoting && !isspace(npb->str[i]))))
 		++i;
 	const size_t lenVal = i - iVal;
 
+	if (i < npb->strLen && npb->str[i] == quoting)
+		++i;
+	else if (quoting)
+		goto done;
+
 	/* parsing OK */
 	*offs = i;
 	r = 0;
-- 
2.21.1

